### PR TITLE
Asynchronous signing

### PR DESCRIPTION
**Pull request by sharksforarms**
For: https://github.com/h2o/h2o/pull/3064

Splits the certificate and certificate verify function and adds a `PTLS_STATE_SERVER_GENERATING_CERTIFICATE_VERIFY` state and `PTLS_ERROR_ASYNC_OPERATION` return code

References:
https://github.com/h2o/picotls/pull/291

### CODE REVIEW

**Review by kazuho — COMMENTED** [review]
Thank you for the changes. Looks good, I only have a few questions.

File: lib/picotls.c [hunk 1]
```diff
@@ -4427,6 +4488,9 @@ void ptls_free(ptls_t *tls)
     if (tls->certificate_verify.cb != NULL) {
         tls->certificate_verify.cb(tls->certificate_verify.verify_ctx, 0, ptls_iovec_init(NULL, 0), ptls_iovec_init(NULL, 0));
     }
+    if (tls->server.sign_certificate.cb != NULL) {
+        tls->server.sign_certificate.cb(tls->server.sign_certificate.sign_certificate_ctx);
+    }
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/picotls.c:4493]: That's correct.

The callback is being set and cleared in the backend implementation:

The cancel callback is set when an async job is started (ASYNC_PAUSE) https://github.com/sharksforarms/picotls/blob/sharksforarms/async/lib/openssl.c#L861 and cleared at the end of the function (on error or job finished) https://github.com/sharksforarms/picotls/blob/sharksforarms/async/lib/openssl.c#L901

File: include/picotls.h [hunk 2]
```diff
@@ -778,6 +785,10 @@ struct st_ptls_context_t {
      * boolean indicating if the cipher-suite should be chosen based on server's preference
      */
     unsigned server_cipher_preference : 1;
+    /**
+     * boolean indicating if handshaking should be asynchronous
+     */
+    unsigned async_handshake : 1;
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply include/picotls.h]: Yes you're correct here, I don't think this flag should be needed in the end. I've added it as a way to temporarily disable async, for example in QUIC context, quicly uses a different code path, which I haven't gotten around to yet.

**Comment by kazuho** [comment]
Subsumed by #422.

**Pull request closed without merging**